### ISSUE TITLE
&hearts; entity not recognized

### ISSUE THREAD

**Issue by gmoore016** [sul-cidr/patent_data_extractor#8]
Hi @simonwiles --

Hope you're doing well! We're running the parser and it's mostly working great, but we're running into a couple issues we're hoping you can take a look at.

We're currently running the old parser with the following configuration file:
```
PATDOC:
  <entity>: patent
  <primary_key>: SDOBI/B200/B210/DNUM 
  <fields>:
    "SDODE/GOVINT/*":
      # the asterisk specifies that all sub-elements of each us-patent-grant/description
      #  will be reduced to their text content, and the results of each such
      #  reduction (for each us-patent-grant/claims/claim element, should there be more
      #  than one) will be concatenated together with the `joiner` string -- in
      #  this case, a newline.
      <fieldname>: description_text
      <joiner>: "|#|"
```

In `pg021112`, we receive the following error message: 

```
Unable to parse XML document ending at line 849185 (enable debugging -v to dump doc to console):
        Entity 'hearts' not defined, line 426, column 373
Warning:__main__:Unable to parse XML document ending at line 849185 (enable debugging -v to dump doc to console): 
        Entity 'hearts' not defined, line 426, column 373
```

Further, in `pg030902` we receive a very similar error:

```Unable to parse XML document ending at line 605562 (enable debugging -v to dump doc to console):
         Entity 'hearts' not defined, line 247, column 270
WARNING:__main__:Unable to parse XML document ending at line 605562 (enable debugging -v to dump doc to console):
         Entity 'hearts' not defined, line 247, column 270
```

Looking at the document dumped to console, it looks as if `&hearts;` is a symbol, like the suit of playing cards. For example, the 2003 patent is for a method of displaying playing cards, and they use the suit symbols in their patent description. 

At the moment we're able to ignore the problem using `--continue-on-error`, but it seems like the record isn't added to our data and we wanted to check in with you about it to see if you had ideas for a simple solution.

**Comment by gmoore016** [comment]
On further inspection, it appears this is just a manifestation of the problem solved by this function (thank you @mayalroy). 
https://github.com/sul-cidr/patent_data_extractor/blob/14fefedd4236a2c1bac0b3ed4a50dec955eb5cef/patent_xml_to_csv.py#L27-L35
Would you recommend just adding "\&hearts;" to this sequence and choosing a new unassigned unicode character?

**Comment by simonwiles** [comment]
> Would you recommend just adding "\&hearts;" to this sequence 

Yes,

> and choosing a new unassigned unicode character?

and no :wink: 

As you've identified, that rather unfortunate kludge of a function addresses the XML entities used in the source material that are not to be found in the assortment of DTDs we have at our disposal.  One option, if "correct" DTDs cannot be found, would be to create them, or directly modify the ones we do have.  Partly in the interest of making the deficiency explicit, and partly just for convenience when I was initially developing, I threw in this function to pre-process the source files and sidestep the problem.  It's an almost deliberately inelegant solution.  With that said, I would indeed suggest just adding to the list as needed, to be honest -- at least as long as it looks like we're going to end up with < 10 of these, as appears to be the case right now.

The unicode codepoints used for the substitution weren't chosen from the PUA at random, though.  For the missing MathML elements, I relied on www.mathmlcentral.com and reference.wolfram.com.  To confirm that the appropriate substitution for `\&hearts;` was as expected, I checked how Google Patents renders them here: https://patents.google.com/patent/US6612926.

I'm going to update the script with the appropriate changes and then I'll close this issue.  Feel free to continue to comment here and/or reopen the issue, though, if you feel it's appropriate.

**Comment by gmoore016** [comment]
Fantastic, thanks so much!

**Comment by simonwiles** [comment]
The plot thickened a little when I looked a little closer.  I've updated the docstring for the relevant function with some more notes -- repeating here:

> 
> Substitute out some undefined entities that appear in the XML
> 
> * IndentingNewLine
> * LeftBracketingBar
> * LeftDoubleBracketingBar
> * RightBracketingBar
> 
>   These seem to be MathML symbols, but the mappings that I've used (deriving from
>     https://reference.wolfram.com/language/ref/character/LeftBracketingBar.html and
>     http://www.mathmlcentral.com/characters/glyphs/LeftBracketingBar.html) point to
>     code points in the PUA of the Unicode BMP -- i.e., they're only going to work
>     with specific fonts.
>   It seems like they should be part of mmlextra
>     (see https://www.w3.org/TR/REC-MathML/chap6/byalpha.html), but they're not in
>     any of the versions (plural!) of this file that I have available, or can find
>     documented online (see, e.g., https://www.w3.org/TR/MathML2/mmlextra.html,
>     https://www.w3.org/2003/entities/mathmldoc/mmlextra.html etc.)
> 
>   The version of mmlextra.ent from here:
>     https://github.com/martinklepsch/patalyze/blob/master/resources/parsedir/mmlextra.ent
>     seems to have what's required?
> 
> ----
> 
> * hearts
> 
>   This entity appears in the detailed description for, e.g., 09489911 (note that this
>   description is not extracted by the parser according to the current version of the
>   config file at config/uspto-grants-0105.yaml).
> 
>   In the same context the suit "diamonds" is represented by \&diams;, and clubs and
>   hearts are presented by \<CUSTOM-CHARACTER\> elements which reference external TIFF
>   files... (sigh).  This is despite the face that the appropriate XML entities
>   (\&hearts;, \&diams;, \&clubs;, and \&spades;) are all defined in the DTDs and .ent
>   files available, but for some reason \&hearts; (and \&hearts; alone) is missing from
>   isopub.ent, which is the file actually invoked by the DTD specified in the XML
>   (double sigh).
> 
>   In some of the DTD files the symbols specified for the suits are the white glpyhs
>   (i.e. &#x2661, &#x2662, &#x2664, and &#x2667 for hearts, diamonds, spades and clubs
>   respectively), and in others they are the black glyphs (i.e. &#x2665, &#x2666,
>   &#x2660, and &#x2663) -- see, e.g.
>   https://en.wikipedia.org/wiki/List_of_Unicode_characters#Miscellaneous_Symbols.
> 
>   I've chose the black variant here, as the black variants are used in isopub.ent
>   -- but note that Google Patents has used the black variants for diamonds
>   (presumably from isopub.ent), the white variant for hearts (coopted from another
>   DTD?), and has dropped the \<CUSTOM-CHARACTER\> elements for spades and clubs (see
> https://patents.google.com/patent/US6612926).

**Comment by simonwiles** [comment]
Closed by 6e257a4f.
